tests/unit/test_project.py

```diff
@@ -0,0 +1,158 @@
+        responses.add(
+            responses.POST,
+            self.input_datasets_url,
+            json=self.post_input_datasets_json,
+            status=204,
+        )
+        responses.add(
+            responses.GET, self.input_datasets_url, json=self.get_input_datasets_json
+        )
+        auth = UsernamePasswordAuth("username", "password")
+        unify = Client(auth)
+        dataset = unify.datasets.by_external_id(self.dataset_external_id)
+        project = unify.projects.by_external_id(self.project_external_id)
+        project.add_source_dataset(dataset)
+        alias = project.api_path + "/inputDatasets"
+        input_datasets = project.client.get(alias).successful().json()
+        self.assertEqual(self.dataset_json, input_datasets)
+
+    @responses.activate
+    def test_project_by_external_id__raises_when_not_found(self):
+        responses.add(responses.GET, self.projects_url, json=[])
+        auth = UsernamePasswordAuth("username", "password")
+        unify = Client(auth)
+        with self.assertRaises(KeyError):
+            unify.projects.by_external_id(self.project_external_id)
+
+    @responses.activate
+    def test_project_by_external_id_succeeds(self):
+        responses.add(responses.GET, self.projects_url, json=self.project_json)
+        auth = UsernamePasswordAuth("username", "password")
+        unify = Client(auth)
+        actual_project = unify.projects.by_external_id(self.project_external_id)
+        self.assertEqual(self.project_json[0], actual_project._data)
+
+    @responses.activate
+    def test_project_attributes_get(self):
+        responses.add(responses.GET, self.projects_url, json=self.project_json)
+        responses.add(
+            responses.GET,
+            self.project_attributes_url,
+            json=self.project_attributes_json,
+        )
+        unify = Client(UsernamePasswordAuth("username", "password"))
+        project = unify.projects.by_external_id(self.project_external_id)
+        attributes = list(project.attributes)
+        self.assertEqual(len(self.project_attributes_json), len(attributes))
+        id_attribute = project.attributes.by_name("id")
+        self.assertEqual(self.project_attributes_json[0]["name"], id_attribute.name)
+
+    @responses.activate
+    def test_project_attributes_post(self):
+        responses.add(responses.GET, self.projects_url, json=self.project_json)
+        responses.add(
+            responses.GET,
+            self.project_attributes_url,
+            json=self.project_attributes_json,
+        )
+        responses.add(
+            responses.POST,
+            self.project_attributes_url,
+            json=self.project_attributes_json[0],
+            status=204,
+        )
+        unify = Client(UsernamePasswordAuth("username", "password"))
+        project = unify.projects.by_external_id(self.project_external_id)
+        # The following will fail if it does the wrong thing
```

Comment:
?

Reply:
"The following will fail if it does the wrong thing"?

Reply:
Yes it is true!